docs/architecture.md

```diff
@@ -0,0 +1,438 @@
+which provide a default conversion from `torch` dialect types to the builtin
+`tensor` type and scalar integer/float types. These are not the right choice for
+all backends, but can be copied and adapted by backends. These files closely
+follow the "Type Conversions the Not-So-Hard Way" talk.
+
+
+## Testing
+
+See
+[development.md](https://github.com/llvm/torch-mlir/blob/main/development.md#testing)
+for more details on running tests.
+
+Torch-MLIR has two types of tests:
+
+1. End-to-end execution tests. These compile and run a program and check the
+   result against the expected output from execution on native Torch. These use
+   a homegrown testing framework (see
+   [framework.py](https://github.com/llvm/torch-mlir/blob/7d4a0d0e2b65c7ce8de19993f3b10ad5344fe32b/python/torch_mlir_e2e_test/torchscript/framework.py#L6))
+   and the test suite lives at `python/torch_mlir_e2e_test/test_suite`.
+
+2. Compiler and Python API unit tests. These use LLVM's `lit` testing framework.
+   For example, these might involve using `torch-mlir-opt` to run a pass and
+   check the output with `FileCheck`. `lit` is flexible enough to unit test
+   various Python pieces, importers, and LTC this way as well.
+
+### Why so much end-to-end testing?
+
+Torch-MLIR places a heavy emphasis on end-to-end testing for the following reasons:
+
+Reason 1: Even if a compiler pass produces the output IR that the author
+expected, that output IR may not correctly implement the semantics of the op.
+This is especially true for complex, often-poorly-specified deep learning
+operators that Torch-MLIR is mainly concerned with. It is critical to run these
+against the source of truth to ensure correct implementation.
+
+Reason 2: There are many patterns in Torch-MLIR's backends that really just
+expand one op into other ops without any real logic. When we started Torch-MLIR,
+we were very religious about always having `.mlir` unit tests even for these
+"macro expansion" patterns, but we found that these tests 1) Never caught a bug
+2) Interfered with refactoring / caused spurious extra work (changing op syntax,
+etc.). There is not much point to having a bunch of tests like this, which are
+basically just rewriting the builder calls in a different syntax:
+
+```
+// MyPass.cpp
+b.create<FooOp>(...)
+b.create<BarOp>(...)
+
+// test.mlir
+// CHECK: foo
+// CHECK: bar
+```
+
+Such a test is simply checking that the implementation of an op is the way it
+is. There is no way to change the implementation while having the test pass. So
+the test is fully redundant with the implementation.
+
+Because of this, many Torch-MLIR patches adding support for new ops have no
+`.mlir` unit tests, and only include an end-to-end test (or more likely,
+multiple to test various facets of the op). We generally make sure that our
```

Comment:
nit: `multiple, `

Or, maybe reword this as: 
`only include end-to-end test(s)`